src/multiformats/address.cljc

```diff
@@ -2,29 +2,232 @@
+  "look up protocol attributes, throws if unknown"
+  [protocol-key]
+  (let [protocol-attrs (get protocol->attrs protocol-key)]
+    (when-not protocol-attrs
+      (throw (ex-info (str "No protocol associated with " protocol-key)
+                      {:protocol protocol-key})))
+    protocol-attrs))
+
+(defn- copy-slice
+  ([src offset len]
+   (let [dst (b/byte-array len)]
+     (b/copy src offset dst 0 len)
+     dst))
+  ([src offset]
+   (copy-slice src offset (- (alength ^bytes src) offset))))
+
+
+(defn- decode-value
+  "decode value using protocol `codec` in `data` starting at `offset`"
+  [codec ^bytes data offset]
+  (if-let [fixed-len (codec/fixed-byte-length codec)]
+    (let [value-bytes (copy-slice data offset fixed-len)]
+      [(codec/bytes->str codec value-bytes) fixed-len])
+    ;; varint encoding
+    (let [[value-len num-read] (varint/read-bytes data offset)
+          value-bytes (copy-slice data (+ offset num-read) value-len)]
+      [(codec/bytes->str codec value-bytes) (+ num-read value-len)])))
+
+
+(defn- protocol-value-seq
+  "lazy sequence of `[protocol-key value]`, where
+  `value` is a string representation of the address protocol
+   value of `nil` if protocol has no value"
+  ([^bytes data]
+   (protocol-value-seq data 0))
+  ([^bytes data offset]
+   (when (< offset (alength data))
+       (lazy-seq
+        (let [[code code-len] (varint/read-bytes data offset)
+              protocol-key (get code->protocol code)
+              {:keys [codec]} (ensure-protocol-attrs protocol-key)
+              val-start (+ offset code-len)
+              [value num-read] (decode-value codec data val-start)]
+          (cons [protocol-key value]
+                (protocol-value-seq data (+ offset code-len num-read))))))))
+
+
+(defn- concat-arrs [& arrs]
+  (let [total-len (reduce + (map alength arrs))
+        dst (b/byte-array total-len)]
+    (loop [arrs arrs offset 0]
+      (when-let [^bytes src (first arrs)]
+        (b/copy src 0 dst offset (alength src))
+        (recur (rest arrs) (+ offset (alength src)))))
+    dst))
+
+(defn- entry-bytes
+  "build byte array representation for protocol/value pair,
+   note that `value` may be nil for no-value protocols.
+
+   entry can either be a single keyword for a no-value protocol, or a
+   pair of [protocol-key value]. You can also pass in [key nil] for a protocol
+   not expecting a value"
+  [entry]
+  (let [protocol-key (if (keyword? entry) entry (first entry))
+        value (when (coll? entry) (second entry))
+        {:keys [code, codec]} (ensure-protocol-attrs protocol-key)
+        code-bytes (varint/encode code)
+        val-bytes (if (nil? value)
+                    (b/byte-array 0)
+                    (codec/str->bytes codec value))
+        val-len-bytes (if (codec/fixed-byte-length codec)
+                        (b/byte-array 0)
+                        (varint/encode (alength ^bytes val-bytes)))]
+    (concat-arrs code-bytes val-len-bytes val-bytes)))
```

Comment:
Feels like this could be more efficient by expanding the conditional tree, instead of allocating zero-length byte arrays. On the other hand, it's also not likely to be on any hot paths anytime soon.